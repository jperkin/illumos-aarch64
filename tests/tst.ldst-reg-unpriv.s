/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/*
 * C3.3.11 Load/store register (unprivileged)
 *
 * STTRB  <Wt>, [<Xn|SP>{, #<simm>}]
 * LDTRB  <Wt>, [<Xn|SP>{, #<simm>}]
 * LDTRSB <Rt>, [<Xn|SP>{, #<simm>}]
 * STTRH  <Wt>, [<Xn|SP>{, #<simm>}]
 * LDTRH  <Wt>, [<Xn|SP>{, #<simm>}]
 * LDTRSH <Rt>, [<Xn|SP>{, #<simm>}]
 * STTR   <Rt>, [<Xn|SP>{, #<simm>}]
 * LDTR   <Rt>, [<Xn|SP>{, #<simm>}]
 * LDTRSW <Xt>, [<Xn|SP>{, #<simm>}]
 */
.macro ldst_reg_unpriv op, wreg=0, xreg=0
	.if \wreg
	\op	w0,  [x0]
	\op	wzr, [sp,  #-1]
	\op	w30, [x30, #-256]
	\op	wzr, [sp,  #255]
	.endif
	.if \xreg
	\op	x0,  [x0]
	\op	xzr, [sp,  #-1]
	\op	x30, [x30, #-256]
	\op	xzr, [sp,  #255]
	.endif
.endm

.text
.globl libdis_test
.type libdis_test, @function
libdis_test:
	ldst_reg_unpriv sttrb  1
	ldst_reg_unpriv ldtrb  1
	ldst_reg_unpriv ldtrsb 1 1
	ldst_reg_unpriv sttrh  1
	ldst_reg_unpriv ldtrh  1
	ldst_reg_unpriv ldtrsh 1 1
	ldst_reg_unpriv sttr   1
	ldst_reg_unpriv ldtr   1
	ldst_reg_unpriv ldtrsw 0 1
.size libdis_test, [.-libdis_test]
