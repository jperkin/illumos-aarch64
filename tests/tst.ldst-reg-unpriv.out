    libdis_test:      00 08 00 38  sttrb w0, [x0]
    libdis_test+0x4:  ff fb 1f 38  sttrb wzr, [sp, #-1]
    libdis_test+0x8:  de 0b 10 38  sttrb w30, [x30, #-256]
    libdis_test+0xc:  ff fb 0f 38  sttrb wzr, [sp, #255]
    libdis_test+0x10: 00 08 40 38  ldtrb w0, [x0]
    libdis_test+0x14: ff fb 5f 38  ldtrb wzr, [sp, #-1]
    libdis_test+0x18: de 0b 50 38  ldtrb w30, [x30, #-256]
    libdis_test+0x1c: ff fb 4f 38  ldtrb wzr, [sp, #255]
    libdis_test+0x20: 00 08 c0 38  ldtrsb w0, [x0]
    libdis_test+0x24: ff fb df 38  ldtrsb wzr, [sp, #-1]
    libdis_test+0x28: de 0b d0 38  ldtrsb w30, [x30, #-256]
    libdis_test+0x2c: ff fb cf 38  ldtrsb wzr, [sp, #255]
    libdis_test+0x30: 00 08 80 38  ldtrsb x0, [x0]
    libdis_test+0x34: ff fb 9f 38  ldtrsb xzr, [sp, #-1]
    libdis_test+0x38: de 0b 90 38  ldtrsb x30, [x30, #-256]
    libdis_test+0x3c: ff fb 8f 38  ldtrsb xzr, [sp, #255]
    libdis_test+0x40: 00 08 00 78  sttrh w0, [x0]
    libdis_test+0x44: ff fb 1f 78  sttrh wzr, [sp, #-1]
    libdis_test+0x48: de 0b 10 78  sttrh w30, [x30, #-256]
    libdis_test+0x4c: ff fb 0f 78  sttrh wzr, [sp, #255]
    libdis_test+0x50: 00 08 40 78  ldtrh w0, [x0]
    libdis_test+0x54: ff fb 5f 78  ldtrh wzr, [sp, #-1]
    libdis_test+0x58: de 0b 50 78  ldtrh w30, [x30, #-256]
    libdis_test+0x5c: ff fb 4f 78  ldtrh wzr, [sp, #255]
    libdis_test+0x60: 00 08 c0 78  ldtrsh w0, [x0]
    libdis_test+0x64: ff fb df 78  ldtrsh wzr, [sp, #-1]
    libdis_test+0x68: de 0b d0 78  ldtrsh w30, [x30, #-256]
    libdis_test+0x6c: ff fb cf 78  ldtrsh wzr, [sp, #255]
    libdis_test+0x70: 00 08 80 78  ldtrsh x0, [x0]
    libdis_test+0x74: ff fb 9f 78  ldtrsh xzr, [sp, #-1]
    libdis_test+0x78: de 0b 90 78  ldtrsh x30, [x30, #-256]
    libdis_test+0x7c: ff fb 8f 78  ldtrsh xzr, [sp, #255]
    libdis_test+0x80: 00 08 00 b8  sttr w0, [x0]
    libdis_test+0x84: ff fb 1f b8  sttr wzr, [sp, #-1]
    libdis_test+0x88: de 0b 10 b8  sttr w30, [x30, #-256]
    libdis_test+0x8c: ff fb 0f b8  sttr wzr, [sp, #255]
    libdis_test+0x90: 00 08 40 b8  ldtr w0, [x0]
    libdis_test+0x94: ff fb 5f b8  ldtr wzr, [sp, #-1]
    libdis_test+0x98: de 0b 50 b8  ldtr w30, [x30, #-256]
    libdis_test+0x9c: ff fb 4f b8  ldtr wzr, [sp, #255]
    libdis_test+0xa0: 00 08 80 b8  ldtrsw x0, [x0]
    libdis_test+0xa4: ff fb 9f b8  ldtrsw xzr, [sp, #-1]
    libdis_test+0xa8: de 0b 90 b8  ldtrsw x30, [x30, #-256]
    libdis_test+0xac: ff fb 8f b8  ldtrsw xzr, [sp, #255]
